bson/src/main/org/bson/codecs/configuration/CodecRegistry.java

```diff
@@ -40,4 +47,24 @@ public interface CodecRegistry extends CodecProvider {
      * @throws CodecConfigurationException if the registry does not contain a codec for the given class.
      */
     <T> Codec<T> get(Class<T> clazz);
+
+    /**
+     * Gets a Codec for the given parameterized class, after resolving any type variables with the given type arguments.
+     *
+     * <p>
+     * The default behavior is to throw a {@link CodecConfigurationException}.  Applications are encouraged to either override this
+     * method in their own implementations of this interface, or else use the factory methods in {@link CodecRegistries}.
+     * </p>
+     *
+     * @param clazz the parameterized class
+     * @param typeArguments the type arguments to apply to the parameterized class
+     * @return a codec for the given class, with the given type parameters resolved
+     * @throws CodecConfigurationException by default, if the implementation does not override this method, or if no codec can be found
+     * for the given class and type arguments.
+     * @param <T> the class type
+     * @since 4.8
+     */
+    default <T> Codec<T> get(Class<T> clazz, List<Type> typeArguments) {
```

Comment:
All the driver-provided implementations override the default implementation.   The only problem is if an application has implemented this interface.  This should be rare, but I've seen a few support cases where people missed the memo about `CodecRegistries`. Since currently it's only records that use this functionality and records are already horribly broken when they have any parameterized types, I think this is ok.  But we'll have to make a decision about whether it's safe to start relying on this for PojoCodec, where we could break existing applications on a driver upgrade.

Reply:
> we should annotate the interface with @Evolving 

Agreed,  though we would need to copy the annotation into either `bson-record-codec` or `bson` to do so.

> and remove the default implementation

You don't buy my argument that it's not really breaking until we start using it in PojoCodec?  I thought that was fairly compelling.

Reply:
> Or we could move @Evolving 
I'd prefer not to add another Maven artifact for such a small thing.  Maybe we should just copy it.

> If this can't happen, could you please help me see why?

Yes, it could happen.  But since it's less likely to actually be the case, the runtime exception still seems ok.

Reply:
Its a breaking change and its not 100% necessary: it can be solved without changes to the codec registry at this time. So should it be added in 4.x rather than 5.x? I'm not sure.

Its only called in 3 places: `ContainerCodecHelper`, `LazyCodec` and `RecordCodec` and implemented in 3 CodecRegistry implementations.

If we are worried about breaking changes then we could always add a new interface and check they type of the codec registry and process from there.

Reply:
> we could always add a new interface and check they type of the codec registry

I think it would amount to the same thing as what's in this PR: a runtime exception when the application actually tries to use the new functionality (currently only in the context of record codec)

Reply:
It has the benefit of not breaking any users code and if not available can defer to the none typed version.

This implementation is really internal detail for the RecordCodec and possibly in the future be shared with the PojoCodec (which already has its own implementation that didnt require any breaking changes).

Reply:
> and if not available can defer to the none typed version

Ah, that was the part I was missing!  Yes, we could fallback to the old behavior.  But I'm not sure if that's a better user experience.  How would they find out that they are doing something wrong and correct it?

Reply:
I gave it a shot by adding a new interface--ParameterizationAwareCodecRegistry--that defines the new method, and everything works fine if I change all the CodecRegistry implementations in the driver to implement this interface and add `instancof` checks where required.  

The problem comes if I then replace one of the built-in registries with one that doesn't implement the interface.  Most of the RecordCodecTest cases start to fail, in two different ways:

1. `CodecConfigurationException: Can not decode to a record with type parameters that has not been parameterized`
2. Incorrect decoding, where the type system is violated (e.g. `Document` instances in a list of type `List<FooRecord>` 

The first type of failure is ok, as it manifests as an exception, similar to what would happen with this PR as is.  But the second seems risky, as an application might have more trouble recognizing that decoding is broken, or diagnosing why it's broken when they do notice it.

Reply:
Interested in how 2 comes about - can you provide an example of how you configured it to fail?

Reply:
I demonstrated it in a separate draft PR: #1011   PTAL there

Reply:
Added another commit that implements the idea that I described [here](https://github.com/mongodb/mongo-java-driver/pull/1011#discussion_r989273403).  PTAL

Reply:
And I just realized that the change is unnecessary.  Since all application-provided registries are _already_ wrapped in an instance of `OverridableUuidRepresentationCodecRegistry`, we're already protected, since the application-provided registry is always treated as a CodecProvider.

Reply:
It would be clearer, for example, if the application were to provide in the settings not a `CodecRegistry` but rather a `List<CodecProvider>`.

Reply:
I'm considering opening a ticket to deprecate `MongoClientSettings#codecRegistry` and adding `MongoClientSettings#codecProvider`, and add a new `CodecProviders` builder class similar to `CodecRegistries`.  Alternatively, could just document that `MongoClientSettings#codecRegistry` is treated only as a `CodecProvider`.  The former seems better but more disruptive, the latter seems less disruptive but more likely to just be ignored. Thoughts?

Reply:
> And I just realized that the change is unnecessary. Since all application-provided registries are already wrapped in an instance of OverridableUuidRepresentationCodecRegistry, we're already protected, since the application-provided registry is always treated as a CodecProvider. 

> I'm considering opening a ticket to deprecate MongoClientSettings#codecRegistry and adding MongoClientSettings#codecProvider, and add a new CodecProviders builder class similar to CodecRegistries. Alternatively, could just document that MongoClientSettings#codecRegistry is treated only as a CodecProvider. The former seems better but more disruptive, the latter seems less disruptive but more likely to just be ignored. Thoughts?

Is this actually true, though? It seems that `com.mongodb.client.internal.MongoClientImpl.createChangeStreamIterable` does not wrap `settings.getCodecRegistry()` into `OverridableUuidRepresentationCodecRegistry`, and `ChangeStreamIterableImpl` uses this user-provided `CodecRegistry` as `CodecRegistry` by calling `CodecRegistry.get(Class<T> clazz)`.

Reply:
I will have to double-check, but this seems like a bug.

Reply:
I thought I finally understood the `CodecProvider`/`CodecRegistry`, and started writing why we don't need `<T> Codec<T> get(Class<T> clazz, CodecRegistry registry)`, and only need `<T> Codec<T> get(Class<T> clazz)`. But while doing so, another revelation came and now I believe I see why `CodecProvider` is needed and is indeed what we should ask from user, while asking users to give us a `CodecRegistry` makes no sense. In other words, I am in favour of the first, more disruptive, solution that you proposed.

Reply:
Opened JAVA-4773 to track the bug in MongoClient#watch

Reply:
Opened JAVA-4774 to track the proposed deprecation in MongoClientSettings

Reply:
Let me know if it's ok to resolve this thread

Reply:
I think this thread can be resolved.